enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,354 @@
+pipelines:
+- inputRefs: [ InputPatchy, InputBulldog ]
+  outputRefs: default
+```
+
+### Implementation Details
+
+#### Elasticsearch version-specific workarounds.
+
+The existing Elasticsearch logging store requires "flattening" of the `kubernetes.labels` map:
+
+* Transform key:value object into array of "NAME=VALUE" strings.
+* Replace '.' with '_' in label names.
+
+This makes label keys difficult to use for a consumer.
+There is no automatic way to reverse the process since '.' and '_' are both legal characters in label names.
+
+The new forwarder will check the version of the Elasticsearch Operator (via labels or annotations, TBD).
+Flattening labels will be enabled if the ELO is identified as a 'legacy' version.
+Other output types, and future ELO versions based on the [pipeline proposal][es-pipeline] will receive labels unmodified.
+
+#### Elasticsearch Indexing <a name="es-notes"></a>
+
+To be efficient, Elasticsearch dynamic indexing needs records to contain a bounded set of key names with consistent value types.
+Inconsistent types for the same key cause type mismatch errors.
+Too many key names (including recursive keys in sub-objects) cause an "index explosion" and performance problems.
+This leads to problems when inconsistent JSON records are stored in the same index.
+
+The current data model uses `viaq_index_name` to select the ES index.
+If `schema` is set, it is copied to `viaq_index_name` instead of a default index name.
+Elasticsearch will then create a separate index for each schema.
+
+Provided log entries are consistent with some fixed schema, it will be possible to search/index custom JSON messages in the *existing* Elasticsearch store implementation.
+
+**Note**: It is up to the user to ensure that logs identified with a schema actually are consistent, otherwise index explosion and type confusion are still possible.
+
+With the *new* [Elasticsearch pipeline proposal][es-pipeline] the store will pre-process records and limit indexing to be more robust.
+* Safe handling of `kubernetes.labels`, so forwarder can turn off flattening/de-dotting behavior.
+* Safe handling of `structured`, default limit on indexing depth to avoid problems with deeply nested JSON.
+* Other possible optimizations and checks, outlined in [Elasticsearch pipeline proposal][es-pipeline]
+* Eventually remove `viaq_index_name` and use `schema` directly.
+
+The CLO will check the version of ELO that is deployed automatically configure the `default` output appropriately.
+
+### Risks and Mitigation
+
+Risks:
+* Changes to data model and output format may impact users.
+* Data model may have other issues to be addressed.
+
+Mitigation:
+* Data model changes are additive to avoid compatibility problems.
+* Need full review of data model.
+
+## Design Details
+
+### Open Questions
+
+* Need sync the [Elasticsearch pipeline proposal][es-pipeline] with this document, update both if needed.
+* What is the ES limitation that requires de-dotting?
```

Comment:
In earlier versions of Elasticsearch the dots in field names may have been interpreted in [different ways](https://discuss.elastic.co/t/is-there-a-way-to-allow-dots-in-the-fieldname/90134/6) (or completely forbidden) and this was [confusing to users](https://discuss.elastic.co/t/how-to-handle-dot-in-field-names-in-es-7-0/176511/8). Thus at certain point Elasticsearch setup clear convention about how dots in field names are interpreted:

```json
{ "foo.bar": true }
```
is interpreted as
```json
{
  "foo": {
    "bar": true
  }
}
```
That is the only way how Elasticsearch users should think about JSON data stored in the store. JSON objects are simply hierarchical.

The point is that dots in field names are not only about how the data is stored but also how it can be queried and how the schema model is constructed. And trying to support dots in field names led to issues and confusion.

One consequence of existing convention is that multi-dots in field names lead to parsing error. For example the following field name is invalid:
```json
{ "foo...bar": true }
```

Reply:
Understood, but that sounds to me like a quoting bug either in Elastcsearch or in the way we use it. JSON syntax rules are clear: `{ "foo.bar": true }` means `{ "foo.bar": true }`, there is no JSON rule that says '.' has special meaning in a field name or value. Now fluentd (and I suspect Elasticsearch) use JSONpath notation, which *does* give '.' a special meaning, but it *also* includes quoting rules to ensure there is no confusion - it sounds like either we or elasticsearch should be generating the quoted `'[foo.bar]'` instead of naively using `foo.bar` in a JSONpath expression. We really shouldn't be applying an unrecoverable transformation to important user data (label names) to work around such a bug, especially since k8s explicitly states that many standard label names DO have '.' in them, since they use DNS domain notation.

Reply:
By "unrecoverable" I mean that `k8s.io/blah` and `k8s_io/blah` are valid and distinct pod labels that will both be mangled to `k8s_io/blah` by our current scheme. If this really is a deeply engrained bug/problem in elasticsearch's query mechanism then we should solve it in the ES pipeline, not on the forwarder. Other JSON users will want to see the original data.

Reply:
Linked this conversation to the open questions section to be resolved during implementation.